src/main/scala/io/iohk/ethereum/db/storage/CachedReferenceCountedStorage.scala

```diff
@@ -0,0 +1,315 @@
+
+import akka.util.ByteString
+import boopickle.Default._
+import com.google.common.cache.RemovalNotification
+import io.iohk.ethereum.db.cache.Cache
+import io.iohk.ethereum.db.storage.NodeStorage.{NodeEncoded, NodeHash}
+import io.iohk.ethereum.mpt.{ByteArraySerializable, NodesKeyValueStorage}
+
+import scala.collection.mutable
+
+
+/**
+  * In-memory pruner - All pruning is done in LRU cache, which means all mpt nodes saved to db, are there permanently.
+  * There two occasions where node is saved to disk:
+  *   1 - When cache becomes full, least recently used nodes are flushed to disk. In normal operation, these nodes
+  *       have already survived several pruning cycles, and still have references pointing at them, which makes them
+  *       unlikely to be pruned in future.
+  *   2 - Every every now and then, cache needs to be flushed to disk to bump up the best block number. It leads to
+  *       saving nodes which were in cache long time and survived many pruning cycles,
+  *       but also some junk nodes from last X Blocks (X - kept history)
+  * There are two supporting data structures which are saved to database after processing each block:
+  *   DeathRow  - List of nodes which reference count drop to 0, and can be potentially deleted in future
+  *   ChangeLog - List of changes to nodes reference counts during processing block. It enables rollbacks of state changes
+  *               made by some block.
+  *
+  * It is something between [[ArchiveNodeStorage]] which saves all nodes even if they would become junk right away, but
+  * is really fast performance wise (only writing data) and [[ReferenceCountNodeStorage]] which tries to clear all junk nodes
+  * but it requires being in sync with db (constant read operations) which hutrs performance.
+  *
+  */
+class CachedReferenceCountedStorage(nodeStorage: NodeStorage,
+                                    cache: Cache[ByteString, HeapEntry],
+                                    changeLog: ChangeLog,
+                                    bn: BigInt) extends NodesKeyValueStorage {
+
+  def get(nodeHash: NodeHash): Option[NodeEncoded] = {
+    cache.get(nodeHash).fold {
+      nodeStorage.get(nodeHash).map(enc => HeapEntry.fromBytes(enc)).map {entry =>
+        cache.put(nodeHash, entry)
+        entry.nodeEncoded
+      }
+    }(entry => Some(entry.nodeEncoded))
+  }
+
+  def update(toRemove: Seq[ByteString], toUpsert: Seq[(ByteString, NodeEncoded)]): NodesKeyValueStorage = {
+    changeLog.withChangelog(bn) { blockChangeLog =>
+      toUpsert.foreach{ case (nodeKey, nodeValue) =>
+        val (updatedValue, change) = {
+          val fromCache = cache.get(nodeKey)
+          if (fromCache.isDefined)
+            (fromCache.get.incrementParents(bn), Increase(nodeKey))
+          else
+            (HeapEntry(nodeValue, 1, bn), New(nodeKey))
+        }
+
+        cache.put(nodeKey, updatedValue)
+        blockChangeLog.registerChange(change, updatedValue.numOfParents)
```

Comment:
i am not sure what you have in mind, add some details ?

Reply:
i was just curious if it's possible to call only once combination of `cache.put` and `blockChangeLog.registerChange` instead of calling it twice (59-60 and 67-68 lines)

Reply:
duplication is only in two places which are near each other and used in simillar context, another foreach/map will only complicate code, so i would leave it as it is